File: day09/mirage.pl
#!/usr/bin/env perl
use v5.38;
use List::Util qw(any sum);

my @reports = map {[split]} <>;

printf "Part 1: %d\n", sum(map {next_val($_)} @reports);
printf "Part 2: %d\n", sum(map {prev_val($_)} @reports);

sub extrapolate($report) {
    my @seqs = ($report);
    push @seqs, [map { $seqs[-1][$_+1] - $seqs[-1][$_] } 0..$seqs[-1]->$#* - 1] while any {$_} $seqs[-1]->@*;
    return \@seqs;
}

sub next_val($report) {
    my $seqs = extrapolate($report);

    my $delta = 0;
    for (my $i = $#$seqs - 1; $i >= 0; $i--) {
	$delta = $seqs->[$i][-1] + $delta;
    }

    return $delta;
}

sub prev_val($report) {
    my $seqs = extrapolate($report);

    my $delta = 0;
    for (my $i = $#$seqs - 1; $i >= 0; $i--) {
	$delta = $seqs->[$i][0] - $delta;
    }

    return $delta;
}

File: day09/mirage.py
import sys

def extrapolate(report):
    seqs = [report]
    while any(seqs[-1]):
        seqs.append([seqs[-1][i+1] - seqs[-1][i] for i in range(len(seqs[-1])-1)])

    return seqs

def next_val(report):
    seqs = extrapolate(report)
    delta = 0
    for i in range(len(seqs)-2, -1, -1):
        delta = seqs[i][-1] + delta

    return delta

def prev_val(report):
    seqs = extrapolate(report)
    delta = 0
    for i in range(len(seqs)-2, -1, -1):
        delta = seqs[i][0] - delta

    return delta

def main():
    with open(sys.argv[1]) as f:
        reports = [list(map(int, line.split())) for line in f]

    print('Part 1:', sum(map(next_val, reports)))
    print('Part 2:', sum(map(prev_val, reports)))

main()
